swarm/metrics.py

```diff
@@ -0,0 +1,165 @@
+    It averages along the all the point predictions in x for a set of predictions.
+    Args:
+        swarm_ypreds: A numpy array of predictions. This can have different numbers of dimensions,
+            eg [bee, epoch, x_step] or [epoch, x_step], so long as the final dimension represents the
+            predictions at different x values, and has same length as y.
+        y: The true values predictions. A 1-d array.
+
+    Returns: A numpy array of dimension one less than the input `swarm_ypreds`, with the evaluated root-mean-squared
+    loss across all predictions.
+    """
+    dims = len(swarm_ypreds.shape)
+    squares = np.square(np.subtract(swarm_ypreds, y))
+    losses = np.apply_along_axis(np.mean, dims-1, squares)
+    return losses
+
+
+def iteration_finder(epoch_scores: [np.ndarray, list],
+                     threshold: float,
+                     threshold_type: str = "nominal",
+                     criterion: str = "first",
+                     direction: str = "below",
+                     reference_iteration: int = 0) -> [int, None]:
+    """
+    This is a function intended to find when a certain score of interest is achieved during a training process.
+    A single vector of scores representing the assessment of the whole swarm at a given epoch/batch (ie iteration)
+    is required.
+
+    In a typical case, you might want to find when a loss is below a suitable threshold, eg .05.
+    An alternative might be that you want to find when the range of standard deviations at all x values has
+    falled below a threshold. (This could indicate knots/nodes eliminated.) Or the median/mean/max standard deviation
+    has fallen below a threshold. (This would indicate a level of coherence within the swarm.)
+
+    The threshold can have two different interpretations, either as a nominal value or a ratio relative to a reference
+    value at a given iteration (often the first).  For different regimes of target functions the nominal evaluation score
+    may vary widely widely in terms of magnitude. So the threshold 0.05 can be used to indicate that the score reaches
+    5% of whatever the score was at the first epoch/batch. For ascending scores, a value greater than one may be
+    helpful. i.e. 20 would indicate improving starting value 20 times over.
+
+    This function also allows you to set a criterion, based on whether it should return the first epoch/batch at which
+    the threshold was crossed, or the first time it was crossed and never thereafter exceeded. (I.e. always remains
+    below that threshold in all tested epochs/batches.)
+    Args:
+        epoch_scores: A list or 1-d Numpy array of floats representing the scores at each epoch/catch.
+        threshold: The nominal value you want the score to first get below or above
+        threshold_type: "ratio" or "nominal". Indicate whether you would like to use the threshold as a nominal value, or a ratio that
```

Comment:
ratio might struggle when you cross the zero line. Most losses are constrained to be positive, but something to consider

Reply:
Hmmm... true. Handling that robustly adds a bit of a fiddle though. Need to rescale.  Actually, no, I think that only an idiot would deliberately choose 'ratio' (which is not default) on a score that crossed over 0.  They can deal with that. Those dealing with sensible losses can have the amenity, the idiots can pick up their own mess.

Reply:
Might be a good idea to do a check and throw a `ValueError`?

Reply:
Good idea

Reply:
Done